NumLib/NamedFunctionCaller.h

```diff
@@ -0,0 +1,133 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef NUMLIB_NAMEDFUNCTIONCALLER_H
+#define NUMLIB_NAMEDFUNCTIONCALLER_H
+
+#include <map>
+#include <vector>
+#include "NamedFunction.h"
+
+namespace NumLib
+{
+class SpecialFunctionCaller;
+
+//! Builds expression trees of named functions dynamically at runtime.
+class NamedFunctionCaller final
+{
+public:
+    //! Constructs an instance whose unbound arguments have the given names.
+    explicit NamedFunctionCaller(
+        std::initializer_list<std::string> unbound_arguments);
+
+    //! Adds the given named function
+    void addNamedFunction(NamedFunction&& fct);
+    std::vector<NamedFunction> const& getNamedFunctions() const
+    {
+        return _named_functions;
+    }
+
+    //! Declares that the argument with name \c sink_args of the function \c
+    //! sink_fct is being computed by the function \c source.
+    //!
+    //! The functions involved need not already be known to the
+    //! NamedFunctionCaller.
+    void plug(std::string const& sink_fct, std::string const& sink_arg,
+              std::string const& source);
+
+    //! Actually plug all the plugs previously declared.
+    //!
+    //! \pre All functions involved must have been added.
+    void applyPlugs();
+
+    //! Calls the function with the given name with the given unbound arguments.
+    //!
+    //! \pre applyPlugs() must have been called before.
```

Comment:
> why not call applyPlugs() inside of this function?

`applyPlugs()` belongs to the initialization of the instance, `call()` belongs to its application. So I rather would keep the two parts separated. Furthermore, the latter is const, the former is not. (Sure, those are not very strong arguments.)

> do we need this function? A user can call it using a result of getSpecialFunction().

It's not strictly needed. I can remove it.